CurvedScalarWave executables and BC

prayush opened the pull request:
## Proposed changes

- Adding executables to evolve the curved scalar wave system in 3D using regular spherical or plane wave in Kerr spacetime as initial data.
- Also adding action to impose Bjorhus type boundary conditions at external boundary

<s>**Note**: The first  commit is cherry-picked from #1832. Please review from 2nd commit onward.</s>

**Note**: The first two commits make up #2600 and #2601 respectively. Please review from 3rd commit onward.


### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

nilsvu commented:
@prayush Do you think you can combine the ScalarWave and CurvedScalarWave executables so you don't have to duplicate the `Metavariables`? Here's an example how I handle a curved background geometry for the elliptic Poisson system: #1910. The idea is to share as much as possible between the systems. You can probably share the ScalarWave Tags, the system and the Metavariables by adding just a few compile-time switches. What do you think? If you prefer to talk in a hangout, just ping me on Slack.

prayush commented:
> @prayush Do you think you can combine the ScalarWave and CurvedScalarWave executables so you don't have to duplicate the `Metavariables`? Here's an example how I handle a curved background geometry for the elliptic Poisson system: #1910. The idea is to share as much as possible between the systems. You can probably share the ScalarWave Tags, the system and the Metavariables by adding just a few compile-time switches. What do you think? If you prefer to talk in a hangout, just ping me on Slack.

Hi @nilsleiffischer - just wanted to note that I haven't forgotten about this comment. I am waiting for all dependencies of this PR to get merged first before resolving this. I like the idea of sharing as much as possible between the systems, but I'll need to think a bit more carefully about how to achieve this with minimal disruption. Thanks!

nilsvu commented:
@prayush ok thanks for letting me know! I'm bringing this up early because things like the numerical flux or the system equations (e.g. #1770, #1835) could also benefit from this treatment, and it's probably easier to review and merge the correct code directly instead of refactoring it later.

prayush committed:
Add executables for the CurvedScalarWave system

  - Exe: EvolveCurvedRegularSphericalWave3D
  - Exe: EvolveCurvedPlaneWave3D
  - Include input file for spherical wave
  - No BCs imposed on external boundary

kidder commented on src/Evolution/Executables/CurvedScalarWave/EvolveCurvedScalarWave.hpp:
```diff
@@ -0,0 +1,377 @@
+
+  using initialization_actions = tmpl::list<
+      Initialization::Actions::TimeAndTimeStep<EvolutionMetavars>,
+      //   dg::Actions::InitializeDomain<volume_dim>,
```
remove

prayush committed:
Add Action to impose Bjorhus BC for CurvedScalarWave system

c.f. Holst et al (2004) and SpEC

prayush committed:
Add ComputeTag to compute Analytic Data

prayush committed:
Correct the compute tag that calculates the largest char speed

The purpose of ComputeLargestCharacteristicSpeed is to estimate
the largest characteristic speed in the bulk, such that the same
can be used to estimate the Courant factor when adjusting timestep
size. The current version could only do that for a given surface
along which normals are known. This commit changes that to get
the bulk maximum characteristic speed instead, which would
correspond to the maximum over all points in the bulk with an
optimally directed normal assumed at each point.

prayush committed:
Update the CurvedScalarWave system:

 - Move time derivatives to their own header/source files
 - Update time derivative computer struct to work with
   dg::Actions::ComputeTimeDerivative
 - Move upwind correction computer struct to separate
   header/source files
 - Update tests accordingly

prayush committed:
Impose Bjorhus BC in CSW exe

nilsdeppe commented:
Should we keep this PR open or can it be closed? It's marked as "duplicate", but maybe it's being used for sharing code :)

nilsdeppe commented:
Was done in #3729
